server/lib/nl/nl_utterance.py

```diff
@@ -0,0 +1,226 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+#
+# Utterance data structure and associate libraries
+#
+
+import logging
+
+from typing import List, Dict
+
+from dataclasses import dataclass
+from enum import IntEnum
+from lib.nl.nl_detection import ContainedInPlaceType, ClassificationType, RankingType, \
+  Detection, NLClassifier, Place, RankingClassificationAttributes, \
+    ContainedInClassificationAttributes, SimpleClassificationAttributes
+
+# How far back does the context go back.
+CNTXT_LOOKBACK_LIMIT = 5
+
+
+# Forward declaration since Utterance contains a pointer to itself.
+class Utterance:
+  pass
+
+
+# Primary charts are about variables/places directly requested by the user.
+# Secondary charts are ones about expansions of the primary variables/places.
+class ChartOriginType(IntEnum):
+  PRIMARY_CHART = 0
+  SECONDARY_CHART = 1
+
+
+# Type of chart.
+class ChartType(IntEnum):
+  TIMELINE_CHART = 0
+  MAP_CHART = 1
+  RANKING_CHART = 2
+  BAR_CHART = 3
+  PLACE_OVERVIEW = 4
+  SCATTER_CHART = 5
+
+
+# Enough of a spec per chart to create the chart config proto.
+@dataclass
+class ChartSpec:
+  chart_type: ChartType
+  utterance: Utterance
+  places: List[Place]
+  svs: List[str]
+  # A list of key-value attributes interpreted per chart_type
+  attr: Dict
+
+
+# The main Utterance data structure that represents all state
+# associated with a user issued query. The past utterances
+# form the context saved in the client and shipped to the server.
+@dataclass
+class Utterance:
+  # Linked list of past utterances
+  prev_utterance: Utterance
+  # Unmodified user-issued query
+  query: str
+  # Result of classification
+  detection: Detection
+  # A characterization of the query
+  query_type: ClassificationType
```

Comment:
is the query_type relevant to the context? For example "tell me about san jose" could be an "overview" query or "query stat var from context"?